### DISCUSSION TITLE
Having trouble switching server from V+ to just individual BepInEx mods like EpicLoot.

### DISCUSSION THREAD

**Discussion by altrae** [lloesche/valheim-server-docker#309] [Q&A]
I've had V+ up and running for a while, but recently decided I want to try running just Epic Loot and some other mods without V+. I set `BEPINEX` to true and `VALHEIM_PLUS` to false, but when I start the server I'm not seeing a new directory created like there was for V+. With `VALHEIM_PLUS` set to true a directory called `valheimplus` is created with the config files as the instructions state, but when I start it with just `BEPINEX` set to true no new directory called `bepinex` or any other is created. I am wondering if I'm just doing something wrong.

**My Environment config:**
![image](https://user-images.githubusercontent.com/5230238/114248917-a9d1b300-995e-11eb-9895-058bf92aee0b.png)

**My Volume config:**
![image](https://user-images.githubusercontent.com/5230238/114248981-db4a7e80-995e-11eb-8ca9-7a47bb8bfd16.png)

**Comment by lloesche** [comment]
That sounds like you did not update the container to the latest version (https://github.com/lloesche/valheim-server-docker#updating-the-container-image-to-the-latest-version).
Like, if it's an old container that does not know about `BEPINEX` yet it would behave exactly as you described.

See if updating the container image helps. If not attach the entire startup log here or in a separate issue.

Also FYI, with the latest container you can now add additional BepInEx plugins into the `/config/valheimplus/plugins` folder. Which would allow you to keep using ValheimPlus and the `VPCFG` config vars you've currently set in addition to any extra mods.
When using BepInEx directly you can theoretically still add the ValheimPlus plugin dll to it but it won't recognize the `VPCFG` settings anymore as those are not generic BepInEx settings.

**Reply by altrae** [reply]
Thanks for your help again. So I've tried updating the container following the linked instructions and it seemed to update, but I am not sure how to see the container version to verify. I did go ahead and update the volume paths from:
File/Folder: docker/Valheim | Mount Path: /config

to

File/Folder: docker/Valheim/config | Mount Path: /config
File/Folder: docker/Valheim/data | Mount Path: /opt/valheim

When the server starts up now I see some new folders under data that I didn't previously, and I have tried restarting the server with different options as well as renaming the valheimplus directory temporarily to make sure it would still be created if V+ was enabled which it does. However, every time I stop the server and switch the flags so that only BEPINEX is true and VALHEIM_PLUS is false it still doesn't create the new bepinex directory as I'm expecting. I'm also not seeing a plugins directory get created in the config/valheimplus directory, only config files.

I've gone ahead and attached the log, and I apologize it's gotten a bit large from multiple restarts while trying different options.

[lloesche-valheim-server.zip](https://github.com/lloesche/valheim-server-docker/files/6291123/lloesche-valheim-server.zip)

**Reply by lloesche** [reply]
From the log it looks like you missed some step in the container update procedure. The container is definitely not the latest version. Current versions output their git commit hash and some filesystem information on startup which is missing in the log. Meaning it must be a version that's a couple of weeks old.
Did you clear the current container? The step where it says "All the data in the current container will be lost".

**Reply by altrae** [reply]
I did clear the container, and that's why the log is only from the last day tho I hadn't updated it for about a month previous. I'll try clearing and wiping the container again and report back.

**Reply by altrae** [reply]
This time instead of just clearing the container, I exported settings, deleted the container and image, then redownloaded and imported using the old settings. The container started and this time it created the config/bepinex properly so that is correct. I do have this error in the beginning of the log tho, but I'm not sure of it's severity because I did get the 'Game server connected' message.

![image](https://user-images.githubusercontent.com/5230238/114294676-d15f7300-9a65-11eb-9b10-0fbd4c25c1e5.png)

**Reply by lloesche** [reply]
Ah yes, so like it says in that message, this is a quirk of Synology NAS. The container defines a startup command: https://github.com/lloesche/valheim-server-docker/blob/3fec572aa9844a90b06dd38f6d05d18b7ff61a43/Dockerfile#L161

which used to be `/usr/bin/supervisord` and has changed a while ago to `/usr/local/sbin/bootstrap`. This startup command is stored inside each container image. For some reason Synology stores that startup CMD independent of the container. So when you update the container it is using the startup CMD of an old version of the container with a new version of the container.

To not break things for Synology users I added a small shim (`fake-supervisord` in the repo) that tells users their container manager is calling the wrong startup CMD and then runs the correct one. Eventually that shim will be deprecated.

The only workaround is to completely delete the container and recreate it from scratch. Or if you have the settings exported and they are stored in some sort of text format you might try to edit them with a text editor and change the string that says `/usr/bin/supervisord` to `/usr/local/sbin/bootstrap`.

The `/config` and `/opt/valheim` volume mounts can be reused of course so all the data is safe and stored independently of the container itself.

As far as your server is concerned all is good. You can play and should now have the latest version. Recreating the container to use the new CMD can be done whenever is convenient or when you're getting too annoyed of that nagging startup message 🙂

**Reply by altrae** [reply]
Oh very nice, I did update the command, recreated the container, and no longer have the message so all is well. Thanks!!

**Comment by MAL0W** [comment]
It does create the bepinex folders for me on my Synology, but when I enter the server, no mods appear.

Let me know if you get it on your Synology friend @altrae

**Reply by altrae** [reply]
Once you have the folder you need to add your mods to the plugins directory which should be at config/bepinex/plugins so it will look something like this:
![image](https://user-images.githubusercontent.com/5230238/114434081-37dcb200-9b88-11eb-89f6-b6871f2c1057.png)

edit: one more thing, once you've added the mods you will need to restart your server for it to find them. You can view the logs at that point and verify they are getting loaded.
![image](https://user-images.githubusercontent.com/5230238/114434558-c4877000-9b88-11eb-8a36-50287442c61d.png)
